pkg/clients/lateinit.go

```diff
@@ -0,0 +1,373 @@
+}
+
+type filterArr []NameFilter
+
+func (fArr filterArr) filter(name string) bool {
+	for _, f := range fArr {
+		if f.filter(name) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// CanonicalNameFilter returns a `NameFilter` option that filters all specified canonical CR field names.
+//   Example: `CanonicalNameFilter("a.b.c", "a.b.d", "a.b.e")`
+func CanonicalNameFilter(cNames ...string) NameFilter {
+	return func(name string) bool {
+		for _, n := range cNames {
+			if name == n {
+				return true
+			}
+		}
+
+		return false
+	}
+}
+
+// NameMapper defines a transformation from CR field names to response field names
+type NameMapper func(string) string
+
+// Apply applies the receiver `NameMapper` to the specified `LateInitOptions`
+func (mapper NameMapper) Apply(options *LateInitOptions) {
+	options.nameMappers = append(options.nameMappers, mapper)
+}
+
+func (mapper NameMapper) getName(name string) string {
+	if mapper == nil {
+		return name
+	}
+
+	return mapper(name)
+}
+
+type mapperArr []NameMapper
+
+func (mArr mapperArr) getName(name string) string {
+	result := name
+
+	for _, m := range mArr {
+		result = m.getName(result)
+	}
+
+	return result
+}
+
+// SuffixReplacer returns a `NameMapper` as a `LateInitOption` that
+//   can be used to replace the specified `suffix` on a CR field name
+//   with the specified `replace` string to obtain the source
+//   response field name.
+//   Example: `SuffixReplacer("ID", "Id")` tells
+//   `LateInitializeFromResponse` to fill a target CR field with name `FieldID`
+//   from a corresponding response field with name `FieldId`.
+func SuffixReplacer(suffix, replace string) NameMapper {
+	return func(s string) string {
+		trimmed := strings.TrimSuffix(s, suffix)
+
+		if trimmed != s {
+			return trimmed + replace
+		}
+
+		return s
+	}
+}
+
+// Replacer returns a `NameMapper` as a `LateInitOption` that
+//   that replaces all occurrences of string `old` to `new` in a
+//   target CR field name to obtain the corresponding
+//   source response field name.
+func Replacer(old, new string) NameMapper {
+	return func(s string) string {
+		return strings.ReplaceAll(s, old, new)
+	}
+}
+
+// MapReplacer returns a `NameMapper` as a `LateInitOption` that
+//   uses the specified `map[string]string` to map from
+//   target CR field names to corresponding source response field names.
+func MapReplacer(m map[string]string) NameMapper {
+	return func(s string) string {
+		if result, ok := m[s]; ok {
+			return result
+		}
+
+		return s
+	}
+}
+
+// LateInitializeFromResponse Copy unset (nil) values from responseObject to crObject
+//   Both crObject and responseObject must be pointers to structs.
+//	 Otherwise, an error will be returned. Returns `true` if at least one field has been stored
+//   from source `responseObject` into a corresponding field of target `crObject`.
+// nolint:gocyclo
+func LateInitializeFromResponse(parentName string, crObject interface{}, responseObject interface{},
```

Comment:
It seems that this library is generic enough to be used by other resources as well but I'm not sure if folks would be comfortable with this level of complexity. From my experience, generalization through loosely typed methods, like reflect or json, leads to bugs that are very hard to debug for inexperienced contributors (and time-consuming for others), which happens once in a while with `CreateJSONPatch` implementation we have for updates. We put exceptions in place per-resource to fix them but this leads to either requiring all cases to be tested for that specific resource or letting customers run into them and report back, which is not an experience we'd like users to have. Additionally, we have not tested for it, but it's in the hot execution path (called in every reconcile), so the performance implication is probably higher than simple `if` conditions.

Because of these reasons, we're relying more on the code generation as much as possible to eliminate these disadvantages of implicit methods as much as possible. https://github.com/aws-controllers-k8s/code-generator/pull/9 is the PR that adds support to ACK for generating the `lateInitialize` functions and https://github.com/aws-controllers-k8s/code-generator/pull/17 adds support for `isUpToDate` check in a strongly typed manner. I'll get back to those PRs in the coming days and we'll hopefully be able to use them for all generated resources.

Reply:
We have synced with @ulucinar offline. The code generated by https://github.com/aws-controllers-k8s/code-generator/pull/9 will supersede this functionality. Until that PR is merged, we'll keep this function private to cloudfront controller and once it's merged, we'll replace it with the generated code.